### ISSUE TITLE
`torch.matmul` misses `mm` fold for `[S, 1, H] @ [H, K]` when the size-1 batch dim has a non-contiguous stride

### ISSUE THREAD

**Issue by cuichenx** [pytorch/pytorch#186148]
## Summary

`torch.matmul` falls back to `aten::bmm` for a 3D x 2D matmul with logical shape:

```text
lhs shape  = [384, 1, 4096]
lhs stride = (4096, 1572864, 1)
rhs shape  = [4096, 4096]
```

The middle dimension has size 1, so its stride should be irrelevant for memory indexing. The LHS can be flattened to `[384, 4096]` as a view, and explicitly doing that dispatches to `aten::mm` and is much faster.

This looks like a missed optimization in `should_fold`. The current stride check rejects this tensor because:

```text
4096 != 1572864 * 1
```

But because the dimension with stride `1572864` has size 1, that stride does not affect the actual addressed elements.

Relevant source:

https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp#L1988-L1998

## Minimal Repro

Create `repro_matmul_bs1_fold.py`:

```python
import statistics

import torch


S = 384
B = 1
H = 4096
DTYPE = torch.bfloat16
DEVICE = "cuda"


def synchronize() -> None:
    torch.cuda.synchronize()


def bench(fn, *, warmup: int = 10, reps: int = 50) -> float:
    for _ in range(warmup):
        fn()
    synchronize()

    times = []
    for _ in range(reps):
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        start.record()
        fn()
        end.record()
        synchronize()
        times.append(start.elapsed_time(end))
    return statistics.median(times)


def print_ops(name, fn) -> None:
    with torch.profiler.profile(
        activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
        record_shapes=True,
    ) as prof:
        fn()
        synchronize()

    interesting_ops = {
        "aten::matmul",
        "aten::mm",
        "aten::bmm",
        "aten::reshape",
        "aten::clone",
        "aten::copy_",
    }

    print(f"\n{name} ops:")
    for event in prof.key_averages(group_by_input_shape=True):
        if event.key in interesting_ops:
            print(event.key, event.input_shapes)


def main() -> None:
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required")

    torch.manual_seed(0)

    # Logical shape [S, B, H], but with a non-contiguous size-1 batch dimension.
    # This produces shape=(384, 1, 4096), stride=(4096, 1572864, 1).
    lhs = torch.randn((B, S, H), device=DEVICE, dtype=DTYPE).transpose(0, 1)
    rhs = torch.randn((H, H), device=DEVICE, dtype=DTYPE)

    flattened_lhs = lhs.reshape(-1, H)

    print("device:", torch.cuda.get_device_name(0))
    print("torch:", torch.__version__)
    print("cuda:", torch.version.cuda)
    print("dtype:", DTYPE)
    print("lhs shape:", tuple(lhs.shape))
    print("lhs stride:", tuple(lhs.stride()))
    print("flattened shape:", tuple(flattened_lhs.shape))
    print("flattened stride:", tuple(flattened_lhs.stride()))
    print(
        "reshape shares storage:",
        flattened_lhs.untyped_storage().data_ptr() == lhs.untyped_storage().data_ptr(),
    )

    def old_matmul() -> torch.Tensor:
        return lhs.matmul(rhs)

    def flat_matmul() -> torch.Tensor:
        return lhs.reshape(-1, H).matmul(rhs).reshape(S, B, H)

    torch.testing.assert_close(
        old_matmul().float(),
        flat_matmul().float(),
        rtol=1e-2,
        atol=1e-2,
    )

    old_ms = bench(old_matmul)
    flat_ms = bench(flat_matmul)

    print(f"\nold matmul median:  {old_ms:.3f} ms")
    print(f"flat matmul median: {flat_ms:.3f} ms")
    print(f"speedup:            {old_ms / flat_ms:.2f}x")

    print_ops("old", old_matmul)
    print_ops("flat", flat_matmul)


if __name__ == "__main__":
    main()
```

Run:

```bash
python repro_matmul_bs1_fold.py
```

## Observed Result

On an NVIDIA H100 with BF16:

```text
device: NVIDIA H100 80GB HBM3
torch: 2.12.0a0+0291f960b6.nv26.04.48445190
cuda: 13.2
dtype: torch.bfloat16
lhs shape: (384, 1, 4096)
lhs stride: (4096, 1572864, 1)
flattened shape: (384, 4096)
flattened stride: (4096, 1)
reshape shares storage: True

old matmul median:  1.697 ms
flat matmul median: 0.031 ms
speedup:            54.52x

old ops:
aten::matmul [[384, 1, 4096], [4096, 4096]]
aten::reshape [[384, 1, 4096], []]
aten::reshape [[384, 4096, 4096], []]
aten::bmm [[384, 1, 4096], [384, 4096, 4096]]

flat ops:
aten::reshape [[384, 1, 4096], []]
aten::matmul [[384, 4096], [4096, 4096]]
aten::mm [[384, 4096], [4096, 4096]]
aten::reshape [[384, 4096], []]
```

## Expected Result

For `[S, 1, H] @ [H, K]`, `torch.matmul` should be able to fold the leading dimensions and dispatch to `aten::mm`, because the size-1 batch dimension does not affect memory indexing and flattening to `[S, H]` can be a view.

In this case, the fallback to batched matmul appears unnecessary and substantially slower.

## Why This Matters

This pattern appears in transformer training in frozen-weight linear backward paths, where `grad_output` has shape `[sequence, batch, hidden]`. For packed or transposed layouts, the batch dimension can be size 1 but still have a non-contiguous stride. In that case, `torch.matmul` may dispatch to a slow batched GEMM even though an equivalent view-based flattening dispatches to a much faster `mm`.

## Environment

Observed with:

```text
GPU: NVIDIA H100 80GB HBM3
dtype: bfloat16
torch: 2.12.0a0+0291f960b6.nv26.04.48445190
CUDA: 13.2
```

### Versions

Collecting environment information...
PyTorch version: 2.12.0a0+0291f960b6.nv26.04.48445190
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar  3 2026, 12:15:18) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1063-nvidia-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.2.78
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version: 535.216.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_tensor_ir.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.21.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.21.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8462Y+
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   32
Socket(s):                            2
Stepping:                             8
Frequency boost:                      enabled
CPU(s) scaling MHz:                   120%
CPU max MHz:                          2801.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5600.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            3 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             128 MiB (64 instances)
L3 cache:                             120 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-31,64-95
NUMA node1 CPU(s):                    32-63,96-127
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==7.3.0
[pip3] intel-openmp==2021.4.0
[pip3] mkl==2021.1.1
[pip3] mkl-devel==2021.1.1
[pip3] mkl-include==2021.1.1
[pip3] mypy==2.1.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] nvidia-cudnn-frontend==1.23.0
[pip3] nvidia-pytriton==0.7.0
[pip3] nvtx==0.2.15
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.1.dev20260509
[pip3] open_clip_torch==3.3.0
[pip3] optree==0.19.0
[pip3] tbb==2021.13.1
[pip3] torch==2.12.0a0+0291f960b6.nv26.4.48445190
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torch_tensorrt==2.12.0a0
[pip3] torchao==0.17.0+git42bcdc49
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.9.0
[pip3] torchtitan==0.2.2+git6d1ff9e6
[pip3] torchvision==0.26.0a0+48956e05.nv26.4.48445190
[pip3] torchx==0.7.0
[pip3] triton==3.6.0+git5d72932fc5.nv26.4
[pip3] triton_kernels==1.0.0+git5d72932fc5.nv26.4
[pip3] tritonclient==2.51.0
[conda] Could not collect

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano
